### PR TITLE
Fixes error message and formats command line output

### LINKED ISSUE #129: Fix failed command execution message
## Description
`Failed to execute 'mvn clean verify'`

it is not failed to execute
it is that this command was executed and  failed

## Acceptance Criteria
- the message says exactly what happened – e.g. : `Command * failed – execution of the following commands skipped: *`
- Specify `exit-code` of *command* in message

### PR DESCRIPTION
**Pull request by akhilbojedla**
## Description

Fixes #129  and #21 partially

## Fixes/Implements: #129 #21 

Successful Execution:
image

Failed Execution:
image

## New dependencies introduced?
- [go-runewidth](https://github.com/mattn/go-runewidth) - Helps to get rune width of strings thus letting us format console messages
- [color](https://github.com/fatih/color) - Helps is generating colorized output to console

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the [changelog draft](https://github.com/gopinath-langote/1build/blob/master/docs/CHANGELOG.md) for corresponding release
- [ ] I have update README (If needed)

### CODE REVIEW

**Comment by gopinath-langote** [comment]
Hello @akhilbojedla ,

Thanks for PR.

As library https://github.com/fatih/color is archived on Github and no more maintained by the author, should we use the different library for color?

WDYT?

**Comment by akhilbojedla** [comment]
@gopinath-langote Yes https://github.com/fatih/color is deprecated. But it is already a complete library with little scope of improvements. I would love to go to https://github.com/logrusorgru/aurora but it doesn't seem to follow semantic versioning thus it doesn't allow me to add it to `go.mod`.

**Comment by akhilbojedla** [comment]
> Hello @akhilbojedla ,
> 
> Thanks for PR.
> 
> As library https://github.com/fatih/color is archived on Github and no more maintained by the author, should we use the different library for color?
> 
> ex:- https://github.com/logrusorgru/aurora
> 
> WDYT?

Hi @gopinath-langote ,

I didn't knew about aurora library. I've replaced https://github.com/fatih/color with https://github.com/logrusorgru/aurora

Let me know if you we need to change anything else.

**Comment by gopinath-langote** [comment]
The command output looks good to me.

Is this possible handle width of `---` based on terminal size?
ex:

```
<img width="1426" alt="Screenshot 2019-09-15 at 21 18 59" src="https://user-images.githubusercontent.com/10210778/64926513-87b59480-d7fe-11e9-85ab-d412d96820d8.png">

Can we make it fit to screen size of it is greater than it?

File: cmd/exec/exec.go [hunk 1]
```diff
@@ -1,69 +1,45 @@
...
-			return
-		}
-	}
-	for _, v := range cmdMap {
-		err = executeAndStopIfFailed(v)
-		if err != nil {
-			utils.Println("\nFailed to execute '" + v + "'\n")
-			return
-		}
+	if executionPlan.hasInit() {
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: For better readability - can we call the method as `executionPlan.hasBefore()` and `executionPlan.hasAfter()` - as we call them in yaml config?

WDYT?

File: cmd/exec/exec.go [hunk 2]
```diff
@@ -77,22 +53,136 @@ func getCommandFromName(config config.OneBuildConfiguration, cmd string) string
...
-		value := getCommandFromName(oneBuildConfiguration, name)
-		if value == "" {
-			utils.Println("No command '" + name + "' found in config file\n")
-			config.PrintConfiguration(oneBuildConfiguration)
-			return nil, errors.New("")
+		executionCommand := getCommandFromName(onebuildConfig, name)
+		if executionCommand == "" {
+			utils.PrintlnErr("Error building exectuion plan. Command \"" + name + "\" not found.")
+			config.PrintConfiguration(onebuildConfig)
+			utils.Exit(1)
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: To follow Unix status code - Let's return `exit code 127` for command not found.

File: cmd/exec/exec.go [hunk 3]
```diff
@@ -77,22 +53,136 @@ func getCommandFromName(config config.OneBuildConfiguration, cmd string) string
...
+		executionPlan.Destroy = &CommandContext{"after", after, bashCommand(sh.NewSession(), after)}
 	}
-	return cmdMap, nil
+
+	return executionPlan
+}
+
+// OneBuildExecutionPlan holds all information for the execution strategy
+type OneBuildExecutionPlan struct {
+	Init     *CommandContext
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: Let's call it as `Before,Command & After`

File: cmd/exec/exec.go [hunk 4]
```diff
@@ -77,22 +53,136 @@ func getCommandFromName(config config.OneBuildConfiguration, cmd string) string
...
-	return sh.NewSession().Command("bash", "-c", command).Run()
+// CommandContext holds all meta-data and required information for execution of a command
+type CommandContext struct {
+	name           string
+	command        string
+	commandSession *sh.Session
+}
+
+func (executionPlan *OneBuildExecutionPlan) hasInit() bool {
+	if executionPlan.Init != nil {
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: return `executionPlan.Init != nil`?

File: cmd/exec/exec.go [hunk 5]
```diff
@@ -77,22 +53,136 @@ func getCommandFromName(config config.OneBuildConfiguration, cmd string) string
...
+		return true
+	}
+	return false
+}
+
+func (executionPlan *OneBuildExecutionPlan) hasDestroy() bool {
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: return `executionPlan.Destroy != nil`?

File: cmd/exec/exec.go [hunk 6]
```diff
@@ -77,22 +53,136 @@ func getCommandFromName(config config.OneBuildConfiguration, cmd string) string
...
+}
+
+func (executionPlan *OneBuildExecutionPlan) hasCommands() bool {
+	if len(executionPlan.Commands) > 0 {
+		return true
+	}
+	return false
+}
+
+func (executionPlan *OneBuildExecutionPlan) print() {
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: Can we use `Utils` to print statements not `fmt` everywhere.

The Idea behind using utils is - we can introduce/change any printing library without changing print statements everywhere.

WDYT?

File: testing/fixtures/execute_cmd_fixtures.go [hunk 7]
```diff
@@ -51,8 +59,7 @@ commands:
   - build: echo building project
 `
 
-	expectedOutput := `No command 'random' found in config file
-
+	expectedOutput := aurora.Red("\nError building exectuion plan. Command \"random\" not found.").Bold().String() + `
```
Comment (gopinath-langote) [inline testing/fixtures/execute_cmd_fixtures.go]: Do you think checking color in test is worth and necessary - I am ok to just test the output and not colors.

WDYT?

File: cmd/exec/exec.go (hunk 1 above)
Comment (akhilbojedla) [reply cmd/exec/exec.go]: Maybe. But when we build an execution plan `init` and `destroy` makes more sense. `before` and `after` can be used more like before and after every command.

WDYT?

**Comment by akhilbojedla** [comment]
@gopinath-langote I've addressed most of your comments but one. I've replied to the remaining one. Let me know what are your thoughts

**Pull request merged**